### PR TITLE
Json log output

### PR DESCRIPTION
**Pull request by bat9r**
Added support of JSON log output

# Def. of Done
- [x] Changes described in changelog.md 📝 
- [x] Automatic tests written 🧪
- [x] Manually tested 🧪
- [x] Good code style
- [x] Necessary migrate instructions added to migrate.md 🚀
- [x] No known issues

### CODE REVIEW

**Review by jonas-zipprick — APPROVED** [review]

File: packages/reply/common/src/ts/log.ts [hunk 1]
```diff
@@ -8,10 +8,37 @@ export enum LogTag {
 /**
  * Formats and prints a message, a status code, and the current timestamp to the command line.
  */
-export const log = (message: string, code: Status = Status.Ok, ...tags: LogTag[]): string => {
+export const log = (
+    stack: string | undefined,
```
Comment (romfrolov) [inline packages/reply/common/src/ts/log.ts:12]: An optional parameter should go after the required parameters.

File: packages/reply/common/src/ts/log.ts [hunk 2]
```diff
@@ -8,10 +8,37 @@ export enum LogTag {
...
+    ): string => {
     const tagsString = tags.map(tag => `, #${tag}`).join('');
     const logString: string = `[${code}, ${new Date().toISOString()}${tagsString}]: ${message}`;
     // eslint-disable-next-line no-console
     console.log(logString);
     return logString;
 };
+
+export const logJson = (
+    stack: string | undefined,
```
Comment (romfrolov) [inline packages/reply/common/src/ts/log.ts:35]: An optional parameter should go after the required parameters.

File: packages/reply/common/src/ts/Reply.ts [hunk 3]
```diff
@@ -223,6 +224,10 @@ export class Reply<T = never, S extends Status = Status> {
         return Reply.errStatus('The reply could not be parsed.');
     }
 
+    public static setLogAsJson (logAsJson : boolean):void {
```
Comment (romfrolov) [inline packages/reply/common/src/ts/Reply.ts:227]:
```suggestion
    public static setLogAsJson (logAsJson : boolean): void {
```

File: packages/reply/common/src/ts/Reply.ts [hunk 4]
```diff
@@ -325,11 +330,11 @@ export class Reply<T = never, S extends Status = Status> {
      */
     public log(...tags: LogTag[]): this {
         if (this.notOk()) {
-            log(this.getErrorMessage(), Status.Error, ...tags);
+            log(this.stack, Reply.logAsJson, this.getErrorMessage(), Status.Error, ...tags);
             // eslint-disable-next-line no-console
-            console.log(this.stack);
+            Reply.logAsJson && console.log(this.stack);
```
Comment (romfrolov) [inline packages/reply/common/src/ts/Reply.ts:335]: 1) Why we only print this when `logAsJson` is set to true?
2) Do we even need this `console.log`?

**Review by romfrolov — CHANGES_REQUESTED** [review]
Requested changes and left some questions.

File: packages/reply/common/src/__tests__/Reply.test.ts [hunk 5]
```diff
@@ -14,6 +14,42 @@ test('logDoesNotThrow', () => {
...
+        '\\"date\\":\\"\\d{4}-\\d\\d-\\d\\dT\\d\\d:\\d\\d:\\d\\d(\\.\\d+)?',
+        '(([+-]\\d\\d:\\d\\d)|Z)?\\",\\"stack\\":\\"[^\\"]*\\",',
+        '\\"message\\":\\"', message, '\\",',
+        '\\"code\\":\\"[^\"]*\\",\\"tags":\\[\\]}'].join(''));
+    const regExp2 = new RegExp(['{',
+        '\\"date\\":\\"\\d{4}-\\d\\d-\\d\\dT\\d\\d:\\d\\d:\\d\\d(\\.\\d+)?',
+        '(([+-]\\d\\d:\\d\\d)|Z)?\\",',
+        '\\"stack\\":\\"[^\\"]*\\",',
+        '\\"message\\":\\"Das isch en anerer Tescht.\\",',
+        '\\"code\\":\\"[^\"]*\\",\\"tags":\\[\\]}'].join(''));
```
Comment (romfrolov) [inline packages/reply/common/src/__tests__/Reply.test.ts:30]: Is this the same regex with a different message? Please use a function to generate it to remove code duplication.

File: packages/reply/common/src/__tests__/Reply.test.ts [hunk 6]
```diff
@@ -14,6 +14,42 @@ test('logDoesNotThrow', () => {
...
+test('plainLogTest', () => {
+    Reply.setLogAsJson(false);
+    const consoleSpy = jest.spyOn(console, 'log');
+    const regExp3 = new RegExp(['\\[Error,',
+        '\\s\\d{4}-\\d\\d-\\d\\dT\\d\\d:\\d\\d:\\d\\d(\\.\\d+)?',
+        '(([+-]\\d\\d:\\d\\d)|Z)?\]:\\s',
+        message].join(''));
+    const regExp4 = new RegExp(['\\[Error,',
+        '\\s\\d{4}-\\d\\d-\\d\\dT\\d\\d:\\d\\d:\\d\\d(\\.\\d+)?',
+        '(([+-]\\d\\d:\\d\\d)|Z)?\]:\\sDas isch en anerer Tescht.'].join(''));
```
Comment (romfrolov) [inline packages/reply/common/src/__tests__/Reply.test.ts:46]: Same as above.

File: packages/reply/common/src/__tests__/Reply.test.ts [hunk 7]
```diff
@@ -14,6 +14,42 @@ test('logDoesNotThrow', () => {
...
+    const regExp2 = new RegExp(['{',
+        '\\"date\\":\\"\\d{4}-\\d\\d-\\d\\dT\\d\\d:\\d\\d:\\d\\d(\\.\\d+)?',
+        '(([+-]\\d\\d:\\d\\d)|Z)?\\",',
+        '\\"stack\\":\\"[^\\"]*\\",',
+        '\\"message\\":\\"Das isch en anerer Tescht.\\",',
+        '\\"code\\":\\"[^\"]*\\",\\"tags":\\[\\]}'].join(''));
+    ERR_STATUS.log();
+    expect(consoleSpy.mock.calls[0][0]).toMatch(regExp1);
+    ERR_STATUS2.log();
+    expect(consoleSpy.mock.calls[2][0]).toMatch(regExp2);
```
Comment (romfrolov) [inline packages/reply/common/src/__tests__/Reply.test.ts:34]: You need to use `consoleSpy.mockRestore()` to restore console.

File: packages/reply/common/src/__tests__/Reply.test.ts [hunk 8]
```diff
@@ -14,6 +14,42 @@ test('logDoesNotThrow', () => {
...
+        '\\s\\d{4}-\\d\\d-\\d\\dT\\d\\d:\\d\\d:\\d\\d(\\.\\d+)?',
+        '(([+-]\\d\\d:\\d\\d)|Z)?\]:\\s',
+        message].join(''));
+    const regExp4 = new RegExp(['\\[Error,',
+        '\\s\\d{4}-\\d\\d-\\d\\dT\\d\\d:\\d\\d:\\d\\d(\\.\\d+)?',
+        '(([+-]\\d\\d:\\d\\d)|Z)?\]:\\sDas isch en anerer Tescht.'].join(''));
+    ERR_STATUS.log();
+    expect(consoleSpy.mock.calls[4][0]).toMatch(regExp3);
+    ERR_STATUS2.log();
+    expect(consoleSpy.mock.calls[5][0]).toMatch(regExp4);
```
Comment (romfrolov) [inline packages/reply/common/src/__tests__/Reply.test.ts:50]: You need to use `consoleSpy.mockRestore()` to restore console.

File: packages/reply/common/src/ts/log.ts (hunk 1 above)
Comment (bat9r) [reply packages/reply/common/src/ts/log.ts]: Resolved

File: packages/reply/common/src/ts/log.ts (hunk 2 above)
Comment (bat9r) [reply packages/reply/common/src/ts/log.ts]: Resolved

File: packages/reply/common/src/__tests__/Reply.test.ts (hunk 5 above)
Comment (bat9r) [reply packages/reply/common/src/__tests__/Reply.test.ts]: Resolve

File: packages/reply/common/src/__tests__/Reply.test.ts (hunk 6 above)
Comment (bat9r) [reply packages/reply/common/src/__tests__/Reply.test.ts]: Resolved

**Review by romfrolov — APPROVED** [review]

**Review by jonas-zipprick — APPROVED** [review]

**Pull request merged**